File: clients/client-java/src/main/java/org/apache/gravitino/client/GenericColumn.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.client;

import com.google.common.collect.Lists;
import java.util.Objects;
import org.apache.gravitino.Audit;
import org.apache.gravitino.MetadataObject;
import org.apache.gravitino.MetadataObjects;
import org.apache.gravitino.exceptions.NoSuchTagException;
import org.apache.gravitino.exceptions.TagAlreadyAssociatedException;
import org.apache.gravitino.rel.Column;
import org.apache.gravitino.rel.expressions.Expression;
import org.apache.gravitino.rel.types.Type;
import org.apache.gravitino.tag.SupportsTags;
import org.apache.gravitino.tag.Tag;

/** Represents a generic column. */
public class GenericColumn implements Column, SupportsTags {

  private final Column internalColumn;

  private final MetadataObjectTagOperations objectTagOperations;

  GenericColumn(
      Column column,
      RESTClient restClient,
      String metalake,
      String catalog,
      String schema,
      String table) {
    this.internalColumn = column;
    MetadataObject columnObject =
        MetadataObjects.of(
            Lists.newArrayList(catalog, schema, table, internalColumn.name()),
            MetadataObject.Type.COLUMN);
    this.objectTagOperations = new MetadataObjectTagOperations(metalake, columnObject, restClient);
  }

  @Override
  public SupportsTags supportsTags() {
    return this;
  }

  @Override
  public String[] listTags() {
    return objectTagOperations.listTags();
  }

  @Override
  public Tag[] listTagsInfo() {
    return objectTagOperations.listTagsInfo();
  }

  @Override
  public Tag getTag(String name) throws NoSuchTagException {
    return objectTagOperations.getTag(name);
  }

  @Override
  public String[] associateTags(String[] tagsToAdd, String[] tagsToRemove)
      throws TagAlreadyAssociatedException {
    return objectTagOperations.associateTags(tagsToAdd, tagsToRemove);
  }

  @Override
  public String name() {
    return internalColumn.name();
  }

  @Override
  public Type dataType() {
    return internalColumn.dataType();
  }

  @Override
  public String comment() {
    return internalColumn.comment();
  }

  @Override
  public boolean nullable() {
    return internalColumn.nullable();
  }

  @Override
  public boolean autoIncrement() {
    return internalColumn.autoIncrement();
  }

  @Override
  public Expression defaultValue() {
    return internalColumn.defaultValue();
  }

  @Override
  public Audit auditInfo() {
    return internalColumn.auditInfo();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof GenericColumn)) {
      return false;
    }

    GenericColumn column = (GenericColumn) obj;
    return Objects.equals(internalColumn, column.internalColumn);
  }

  @Override
  public int hashCode() {
    return internalColumn.hashCode();
  }
}

File: clients/client-python/gravitino/client/generic_column.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from typing import Optional, cast

from gravitino.api.metadata_object import MetadataObject
from gravitino.api.metadata_objects import MetadataObjects
from gravitino.api.rel.column import Column
from gravitino.api.rel.expressions.expression import Expression
from gravitino.api.rel.types.type import Type
from gravitino.api.tag.supports_tags import SupportsTags
from gravitino.api.tag.tag import Tag
from gravitino.client.metadata_object_tag_operations import MetadataObjectTagOperations
from gravitino.utils import HTTPClient


class GenericColumn(Column, SupportsTags):
    """Represents a generic column."""

    def __init__(
        self,
        column: Column,
        rest_client: HTTPClient,
        metalake: str,
        catalog: str,
        schema: str,
        table: str,
    ):
        self._internal_column = column
        column_object = MetadataObjects.of(
            [catalog, schema, table, column.name()],
            MetadataObject.Type.COLUMN,
        )
        self.object_tag_operations = MetadataObjectTagOperations(
            metalake,
            column_object,
            rest_client,
        )

    def name(self) -> str:
        return self._internal_column.name()

    def data_type(self) -> Type:
        return self._internal_column.data_type()

    def comment(self) -> Optional[str]:
        return self._internal_column.comment()

    def nullable(self) -> bool:
        return self._internal_column.nullable()

    def auto_increment(self) -> bool:
        return self._internal_column.auto_increment()

    def default_value(self) -> Expression:
        return self._internal_column.default_value()

    def __hash__(self) -> int:
        return hash(self._internal_column)

    def __eq__(self, value: object) -> bool:
        if not isinstance(value, GenericColumn):
            return False
        column = cast(GenericColumn, value)
        return self._internal_column == column._internal_column

    def supports_tags(self) -> SupportsTags:
        return self

    def list_tags(self) -> list[str]:
        return self.object_tag_operations.list_tags()

    def list_tags_info(self) -> list[Tag]:
        return self.object_tag_operations.list_tags_info()

    def get_tag(self, name: str) -> Tag:
        return self.object_tag_operations.get_tag(name)

    def associate_tags(
        self, tags_to_add: list[str], tags_to_remove: list[str]
    ) -> list[str]:
        return self.object_tag_operations.associate_tags(tags_to_add, tags_to_remove)
